README.rst
=======================
``gs.group.properties``
=======================
~~~~~~~~~~~~~~~~~~~~~~~~~~~
Change the group properties
~~~~~~~~~~~~~~~~~~~~~~~~~~~

:Author: `Michael JasonSmith`_
:Contact: Michael JasonSmith <mpj17@onlinegroups.net>
:Date: 2015-02-19
:Organization: `GroupServer.org`_
:Copyright: This document is licensed under a
  `Creative Commons Attribution-Share Alike 4.0 International License`_
  by `OnlineGroups.net`_.

..  _Creative Commons Attribution-Share Alike 4.0 International License:
    http://creativecommons.org/licenses/by-sa/4.0/

Introduction
============

This product provides the basic `Change properties`_ form, and
the `Properties viewlet manager`_.

Change Properties
=================

The *Change properties* form, ``changeproperties.html`` in the
Group context, provides a page for changing the basic group
properties (such as the name).

Properties Viewlet Manager
==========================

The *Properties* viewlet manager provides a list of links to the
*site* administrator, to allow him or her to change the group
properties:

.. code-block:: xml

  <browser:viewlet 
    name="gs-group-messages-ratelimit-home-admin-tab"
    manager="gs.group.properties.interfaces.IGroupAdminPropertiesLinks"
    template="browser/templates/adminlist.pt"
    class="gs.group.member.viewlet.GroupAdminViewlet"
    permission="zope2.ManageUsers"
    weight="5"
    title="Change the posting rate" />

Resources
=========

- Code repository:
  https://github.com/groupserver/gs.group.properties
- Translations:
  https://www.transifex.com/projects/p/gs-group-properties/
- Questions and comments to
  http://groupserver.org/groups/development
- Report bugs at https://redmine.iopen.net/projects/groupserver

.. _GroupServer: http://groupserver.org/
.. _GroupServer.org: http://groupserver.org/
.. _OnlineGroups.Net: https://onlinegroups.net
.. _Michael JasonSmith: http://groupserver.org/p/mpj17

..  LocalWords:  changeproperties html


docs/HISTORY.rst
Changelog
=========

2.5.1 (2016-01-11)
------------------

* [DE] Further updates to the German translation, thanks to
  Cousin Clara, and `Stephan G. Blendinger`_

.. _Stephan G. Blendinger:
   https://www.transifex.com/user/profile/stephanblendinger/


2.5.0 (2015-07-07)
------------------

* Adding the ``Reply-to`` option, closing `Feature 4051`_

.. _Feature 4051: https://redmine.iopen.net/issues/4051

2.4.3 (2015-03-20)
------------------

* Closing an anchor-tag in the English translation of the
  *Properties* page

2.4.2 (2015-03-11)
------------------

* [FR] Adding a French translation, thanks to `Razique Mahroua`_

.. _Razique Mahroua:
   https://www.transifex.com/accounts/profile/Razique/

2.4.1 (2015-02-26)
------------------

* Adding Transifex_ support
* Added German (thanks, Alice_)

.. _Transifex:
   https://www.transifex.com/projects/p/gs-group-properties/
.. _Alice: http://groupserver.org/p/alice

2.4.0 (2014-10-23)
------------------

* Added internationalisation support
* Fixed `Bug 640<https://redmine.iopen.net/issues/640>`_

2.3.1 (2014-10-10)
------------------

* Code cleanup
* Pointing at GitHub as the primary code repository

2.3.0 (2014-07-11)
------------------

* Following ``gs.content.form`` to ``gs.content.form.base``

2.2.1 (2014-03-26)
------------------

* Switching to Unicode literals and absolute import

2.2.0 (2013-07-12)
------------------

* Adding a breadcrumb trail to the page
* Changing which fields are required

2.1.0 (2013-03-25)
------------------

* Following the group-member viewlet
* Code cleanup

2.0.1 (2013-02-20)
------------------

* Using an ``<h4>`` for the label
* Adding some WAI-ARIA properties to the *Properties* list

2.0.0 (2013-01-22)
------------------

* Updating to the new UI
* Moving the *Properties* list to the new *Admin* viewlet

1.2.0 (2012-07-05)
------------------

* Removing the context-menu from the *Change properties* page

1.1.0 (2011-06-27)
------------------

* Creating a viewlet to contain all the *Change group properties*
  links

1.0.0 (2010-11-22)
------------------

* Initial release

..  LocalWords:  Changelog Transifex


version.py
# -*- coding: utf-8 -*-
version = '2.5.1'
release = False

#--------------------------------------------------------------------------#
import sys
if (sys.version_info < (3, )):
    from commands import getstatusoutput
else:
    from subprocess import getstatusoutput  # lint:ok
import datetime
import os
import glob


class CommandError(Exception):
    pass


def execute_command(commandstring):
    status, output = getstatusoutput(commandstring)
    if status != 0:
        m = 'Command "{0}" exited with status {1}'
        msg = m.format(commandstring, status)
        raise CommandError(msg)
    return output


def parse_version_from_package():
    try:
        pkginfo = os.path.join(glob.glob('*.egg-info')[0],
                               'PKG-INFO')
    except:
        pkginfo = ''

    version_string = ''
    if os.path.exists(pkginfo):
        for line in open(pkginfo):
            if line.find('Version: ') == 0:
                version_string = line.strip().split('Version: ')[1].strip()
        if not version_string:
            version_string = '%s-dev' % version
    else:
        version_string = version

    return version_string


def get_version():
    try:
        globalid = execute_command("hg identify -i").strip('+')
        c = "hg log -r %s --template '{date|isodatesec}'" % globalid
        commitdate = execute_command(c)
        # convert date to UTC unix timestamp, using the date command because
        # python date libraries do not stabilise till about 2.6
        dateCommand = 'date -d"%s" --utc +%%s' % commitdate
        timestamp = int(execute_command(dateCommand))

        # finally we have something we can use!
        dt = datetime.datetime.utcfromtimestamp(timestamp)
        datestring = dt.strftime('%Y%m%d%H%M%S')
        if release:
            version_string = version
        else:
            version_string = "%s.dev%s" % (version, datestring)

    except (CommandError, ValueError, TypeError):
        # --=mpj17=-- Usually because we are building out a source-egg,
        # rather than from a Hg source-directory.
        version_string = parse_version_from_package()

    return version_string

if __name__ == '__main__':
    import sys
    sys.stdout.write('{0}\n'.format(get_version()))


gs/group/properties/edit.py
# coding=utf-8
from zope.formlib import form
from Products.Five.browser.pagetemplatefile import ZopeTwoPageTemplateFile
from gs.group.base.form import GroupForm
from gs.content.form.utils import enforce_schema
from interfaces import IGroupProperties

class EditPropertiesForm(GroupForm):
    pageTemplateFileName = 'browser/templates/editproperties.pt'
    template = ZopeTwoPageTemplateFile(pageTemplateFileName)
    
    def __init__(self, context, request):
        GroupForm.__init__(self, context, request)
        enforce_schema(self.context, IGroupProperties)
        self.form_fields = form.Fields(IGroupProperties, render_context=True)
        self.label = u'Edit Group Properties'
        
    @form.action(label=u'Change', failure='handle_change_action_failure')
    def handle_invite(self, action, data):
        form.applyChanges(self.context, self.form_fields, data)
        
        #auditor = Auditor(self.siteInfo, self.groupInfo)
        #admin = createObject('groupserver.LoggedInUser', self.context)
        #auditor.info(CHANGE_ABOUT, admin, '%d' % len(data['aboutText']))
        
        self.status = u'Changed the properties for for '\
          u'<a href="%s">%s</a> has been changed.' %\
          (self.groupInfo.relative_url(), self.groupInfo.name)
          
    def handle_change_action_failure(self, action, data, errors):
        if len(errors) == 1:
            self.status = u'<p>There is an error:</p>'
        else:
            self.status = u'<p>There are errors:</p>'



gs/group/properties/interfaces.py
# -*- coding: utf-8 -*-
############################################################################
#
# Copyright © 2013, 2014, 2015 OnlineGroups.net and Contributors.
# All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
############################################################################
from __future__ import absolute_import, unicode_literals
from zope.interface import Interface
from zope.schema import TextLine, Choice
from zope.schema.vocabulary import SimpleVocabulary, SimpleTerm
from zope.viewlet.interfaces import IViewletManager
from . import GSMessageFactory as _


replyToVocab = SimpleVocabulary([
    SimpleTerm(
        'group', 'group',
        _('reply-to-group', 'Group: replies go to the group.')),
    SimpleTerm(
        'sender', 'sender',
        _('reply-to-sender', 'Author: replies just go to the author of the post.')),
    SimpleTerm(
        'both', 'both',
        _('reply-to-both',
          'Both: replies go to the group, with an extra email to the author.')), ])


class IGroupProperties(Interface):
    title = TextLine(
        title=_('form-label-group-name', 'Group Name'),
        description=_('form-label-group-name-help',
                      'The name of the group'),
        required=True)

    description = TextLine(
        title=_('form-label-description', 'Short Description'),
        description=_('form-label-description-help',
                      'Who and what the online group is for. Shown on the '
                      'Groups page and the Group Home Page.'),
        required=False)

    short_name = TextLine(
        title=_('form-label-subject-line-prefix', 'Subject Line Prefix'),
        description=_('form-label-subject-line-prefix-help',
                      'Shown in the subject line of posts delivered via '
                      'email.'),
        required=True)

    replyTo = Choice(
        title=_('reply-to', 'Reply-to'),
        description=_('This setting determines the behaviour of the Reply-to email header.'),
        vocabulary=replyToVocab,
        required=True,
        default='group')

    mshipCriterion = TextLine(
        title=_('form-label-membership-criterion', 'Membership Criterion'),
        description=_('form-label-membership-criterion-help',
                      "Who is eligible to join the group in the future. "
                      "Displayed only when the group's privacy setting is "
                      '"private" or "secret".'),
        required=False)


class IGroupAdminPropertiesLinks(IViewletManager):
    '''A viewlet manager for the Properties tabs used by the group admin'''


gs/group/properties/change.py
# -*- coding: utf-8 -*-
############################################################################
#
# Copyright © 2013, 2014, 2015 OnlineGroups.net and Contributors.
# All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
############################################################################
from __future__ import absolute_import, unicode_literals, print_function
from zope.cachedescriptors.property import Lazy
from zope.component import createObject
from zope.formlib import form
from Products.Five.browser.pagetemplatefile import ZopeTwoPageTemplateFile
from gs.content.form.base import disabled_text_widget, radio_widget
from gs.content.form.base.utils import enforce_schema
from gs.group.privacy import GroupVisibility
from gs.group.base import GroupForm
from .interfaces import IGroupProperties
from . import GSMessageFactory as _


class ChangePropertiesForm(GroupForm):
    pageTemplateFileName = 'browser/templates/changeproperties.pt'
    template = ZopeTwoPageTemplateFile(pageTemplateFileName)

    def __init__(self, context, request):
        GroupForm.__init__(self, context, request)
        enforce_schema(self.context, IGroupProperties)
        self.groupVisibility = GroupVisibility(self.groupInfo)
        self.label = _('change-properties-page-title',
                       'Change group properties')

    @Lazy
    def mailingList(self):
        retval = createObject('groupserver.MailingListInfo',
                              self.context, self.groupInfo.id)
        return retval

    @staticmethod
    def check_replyTo(v):
        if v not in ('group', 'sender', 'both'):
            raise ValueError('Reply-to can be "group", "sender", or '
                             '"both", not "{0}"'.format(v))

    @Lazy
    def defaultReplyTo(self):
        'The default reply-to, from the ListManager object'
        listManager = self.mailingList.mlist.ListManager
        retval = listManager.getProperty(b'replyto', 'group')
        return retval

    @property
    def replyTo(self):
        'The reply-to property for a group'
        # --=mpj17=-- All this replyTo property stuff is great,
        # but it really needs to be moved to gs.group.list.base
        # at some stage.
        r = self.mailingList.get_property(b'replyto', self.defaultReplyTo)
        retval = 'group' if r is None else r
        retval = 'group' if retval == 'list' else retval  # For old GS installs
        self.check_replyTo(retval)
        return retval

    @replyTo.setter
    def replyTo(self, val):
        self.check_replyTo(val)
        # --=mpj17=-- Aquisition has a bad rep because of things like this aq_explicit
        if hasattr(self.mailingList.mlist.aq_explicit, b'replyto'):
            self.mailingList.mlist.manage_changeProperties(replyto=val)
        else:
            self.mailingList.mlist.manage_addProperty(b'replyto', val, b'string')

    @Lazy
    def form_fields(self):
        retval = form.Fields(IGroupProperties, render_context=True)
        if self.groupVisibility.isPublic:
            retval['mshipCriterion'].custom_widget = disabled_text_widget
        retval['replyTo'].custom_widget = radio_widget
        return retval

    def setUpWidgets(self, ignore_request=False):
        data = {}
        data['replyTo'] = self.replyTo

        self.widgets = form.setUpWidgets(
            self.form_fields, self.prefix, self.context,
            self.request, form=self, data=data,
            ignore_request=ignore_request)
        if self.groupVisibility.isPublic:
            self.widgets['mshipCriterion'].required = False

    @form.action(name='change', label=_('change-button', 'Change'),
                 failure='handle_change_action_failure')
    def handle_change(self, action, data):

        if data['replyTo'] != self.replyTo:
            self.replyTo = data['replyTo']
            del(data['replyTo'])

        form.applyChanges(self.context, self.form_fields, data)
        self.set_subject_prefix(data['short_name'])
        groupLink = '<a href="{groupUrl}">{groupName}</a>.'.format(
            groupUrl=self.groupInfo.relative_url(),
            groupName=self.groupInfo.name)
        self.status = _('changed-status',
                        'Changed the properties of ${groupLink}.',
                        mapping={'groupLink': groupLink})

    def handle_change_action_failure(self, action, data, errors):
        if len(errors) == 1:
            s = 'There is an error:'
        else:
            s = 'There are errors:'
        self.status = '<p>{0}</p>'.format(s)

    def set_subject_prefix(self, prefix):
        '''Set the subject-line prefix by setting the title of the
        mailing list'''
        # Closes https://projects.iopen.net/groupserver/ticket/640
        self.mailingList.mlist.manage_changeProperties(title=prefix)
